src/Runner.Worker/Container/ContainerHooks/ContainerHookManager.cs

```diff
@@ -0,0 +1,260 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using GitHub.DistributedTask.Pipelines.ContextData;
+using GitHub.DistributedTask.WebApi;
+using GitHub.Runner.Common;
+using GitHub.Runner.Common.Util;
+using GitHub.Runner.Sdk;
+using GitHub.Runner.Worker.Handlers;
+using GitHub.Services.WebApi;
+using Newtonsoft.Json.Linq;
+
+namespace GitHub.Runner.Worker.Container.ContainerHooks
+{
+    [ServiceLocator(Default = typeof(ContainerHookManager))]
+    public interface IContainerHookManager : IRunnerService
+    {
+        Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers);
+        Task CleanupJobAsync(IExecutionContext context, List<ContainerInfo> containers);
+        Task ContainerStepAsync(IExecutionContext context, ContainerInfo container);
+        Task ScriptStepAsync(IExecutionContext context, ContainerInfo container, string arguments, string fileName, IDictionary<string, string> environment, string prependPath, string workingDirectory);
+    }
+
+    public class ContainerHookManager : RunnerService, IContainerHookManager
+    {
+        private const string ResponseFolderName = "_hook_responses";
+        private string HookIndexPath;
+
+        public override void Initialize(IHostContext hostContext)
+        {
+            base.Initialize(hostContext);
+            HookIndexPath = $"{Environment.GetEnvironmentVariable(Constants.Hooks.ContainerHooksPath)}";
+        }
+
+        public async Task PrepareJobAsync(IExecutionContext context, List<ContainerInfo> containers)
+        {
+            Trace.Entering();
+            var responsePath = GenerateResponsePath();
+            var jobContainer = containers.Where(c => c.IsJobContainer).FirstOrDefault();
+            var serviceContainers = containers.Where(c => c.IsJobContainer == false).ToList();
+
+            var input = new HookInput
+            {
+                Command = HookCommand.PrepareJob,
+                ResponseFile = responsePath,
+                Args = new PrepareJobArgs
+                {
+                    Container = jobContainer.GetHookContainer(),
+                    Services = serviceContainers.Select(c => c.GetHookContainer()).ToList(),
+                }
+            };
+
+            var prependPath = GetPrependPath(context);
+            var response = await ExecuteHookScript(context, input, ActionRunStage.Pre, prependPath);
+            if (response == null)
+            {
+                return;
+            }
+
+            jobContainer.IsAlpine = (bool) response.IsAlpine;
```

Comment:
Do we need to try and evaluate this some how? Why the cast?

Reply:
`response.IsAlpine` is nullable `bool?` so that after deserialising it we can throw if it's not present. 
`jobContainer.IsAlpine` is non-nullable `bool` because it's mandatory.

I'll replace it with `response.IsAlpine.Value` as at this point IsAlpine is guaranteed to be there.